<a href="https://colab.research.google.com/github/yustinachang/1121_20009_Digital-Image-Processing-and-Analysis/blob/main/HW5/231027_yolov4_tiny_training_1027.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Yolov4-tiny (darknet) 自定義資料集訓練範例  
歐尼克斯實境互動工作室 OmniXRI Jack, 2021.5.28  
參考來源：https://github.com/AlexeyAB/darknet  

#1. 驗證Nvidia GPU及CUDA版本
確認運行環境（本步驟可略過）  
GPU每次配置可能不同，可能為 K80, T4, P100，記憶體大小可能會影響到後續訓練時Patch Size大小。  
若想要好一點的GPU則重覆斷開連線(Ctrl+M)再重新連線，直到出現期望的GPU出現，再進行下一個步驟。  

In [ ]:
!nvidia-smi
!/usr/local/cuda/bin/nvcc --version

Sat Oct 28 09:32:17 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   52C    P8    10W /  70W |      0MiB / 15360MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

#2. 連接Google Drvie
請先在自己的Google Drive上建立一個yolov4-tiny的空檔案夾，方便後續訓練時可暫存權重檔到此處。
掛載Google Drvie（雲端硬碟），點擊網址(Go to this URL in a browser)，允許連結，再複製授權碼貼到空格（Enter your authorization code）中。  
建立捷徑 /my_drive 指向已掛載之Google Drvie (/content/drive/MyDrive/)

In [ ]:
from google.colab import drive

drive.mount('/content/drive')
!ln -s /content/drive/MyDrive/ /my_drive
!ls
!ls /my_drive

Mounted at /content/drive
drive  sample_data
'Colab Notebooks'   MyDrive   yolov4-tiny	 動物倫理   網頁設計
 Dungeon	    other     yolov4-tiny_1027	 筆記	    課程表


#3. 下載darknet及Yolov4-tiny預訓練權重檔

In [ ]:
# 下載darknet
!git clone https://github.com/AlexeyAB/darknet
# 下載yolov4-tiny預訓練權重檔(23.1MB)，僅供測試darknet用，後續用不到
#!wget -N https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v4_pre/yolov4-tiny.weights
!ls

Cloning into 'darknet'...
remote: Enumerating objects: 15825, done.
remote: Counting objects: 100% (286/286), done.
remote: Compressing objects: 100% (144/144), done.
remote: Total 15825 (delta 151), reused 222 (delta 140), pack-reused 15539
Receiving objects: 100% (15825/15825), 14.75 MiB | 21.64 MiB/s, done.
Resolving deltas: 100% (10581/10581), done.
darknet  drive	sample_data


#4. 修改Makefile參數
將GPU, CUDNN, CUDNN_HALF, OPENCV設為可用  
GPU=0 --> GPU=1  
CUDNN=0 --> CUDNN=1  
CUDNN_HALF=0 --> CUDNN_HALF=1  
OPENCV=0 --> OPENCV=1

In [ ]:
%cd darknet
!sed -i 's/GPU=0/GPU=1/' Makefile
!sed -i 's/CUDNN=0/CUDNN=1/' Makefile
!sed -i 's/CUDNN_HALF=0/CUDNN_HALF=1/' Makefile
!sed -i 's/OPENCV=0/OPENCV=1/' Makefile

/content/darknet


#5. 編譯darknet

In [ ]:
!make

mkdir -p ./obj/
mkdir -p backup
mkdir -p results
chmod +x *.sh
g++ -std=c++11 -std=c++11 -Iinclude/ -I3rdparty/stb/include -DOPENCV `pkg-config --cflags opencv4 2> /dev/null || pkg-config --cflags opencv` -DGPU -I/usr/local/cuda/include/ -DCUDNN -DCUDNN_HALF -Wall -Wfatal-errors -Wno-unused-result -Wno-unknown-pragmas -fPIC -rdynamic -Ofast -DOPENCV -DGPU -DCUDNN -I/usr/local/cudnn/include -DCUDNN_HALF -c ./src/image_opencv.cpp -o obj/image_opencv.o
./src/image_opencv.cpp: In function ‘void draw_detections_cv_v3(void**, detection*, int, float, char**, image**, int, int)’:
./src/image_opencv.cpp:945:23: warning: variable ‘rgb’ set but not used []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable-Wunused-but-set-variable]8;;]
  945 |                 float rgb[3];
      |                       ^~~
./src/image_opencv.cpp: In function ‘void cv_draw_object(image, float*, int, int, int*, float*, int*, int, char**)’:
./src/image_opencv.cpp:1443:14: wa

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


#6. 測試darknet編譯結果
1.   檢查目前工作路徑是否在 darknet 路徑
2.   執行darknet物件偵測推論，指定 data, config, weight, image 所在位置(darknet/data/ 路徑下有幾張jpg影像可供測試)。
3.   顯示結果影像predictions.jpg

In [ ]:
!ls
!./darknet  detector test cfg/coco.data cfg/yolov4-tiny.cfg ../yolov4-tiny.weights data/dog.jpg

import cv2 # 導入OpenCV函式庫
from google.colab.patches import cv2_imshow # 導入Colab.patches函式庫

imgResult = cv2.imread('predictions.jpg') # 讀入結果影像
cv2_imshow(imgResult) # 顯示結果影像

3rdparty   CMakeLists.txt	   docker-compose.yml	  LICENSE	 results
backup	   darknet		   Dockerfile.cpu	  Makefile	 scripts
bad.list   DarknetConfig.cmake.in  Dockerfile.gpu	  net_cam_v3.sh  src
build	   darknet_images.py	   image_yolov3.sh	  net_cam_v4.sh  vcpkg.json
build.ps1  darknet.py		   image_yolov4.sh	  obj		 vcpkg.json.opencv23
cfg	   darknet_video.py	   include		  package.xml	 video_yolov3.sh
cmake	   data			   json_mjpeg_streams.sh  README.md	 video_yolov4.sh
 CUDA-version: 11080 (12000), cuDNN: 8.9.0, CUDNN_HALF=1, GPU count: 1  
 CUDNN_HALF=1 
 OpenCV version: 4.5.4
 0 : compute_capability = 750, cudnn_half = 1, GPU: Tesla T4 
net.optimized_memory = 0 
mini_batch = 1, batch = 1, time_steps = 1, train = 0 
   layer   filters  size/strd(dil)      input                output
   0 Create CUDA-stream - 0 
 Create cudnn-handle 0 
conv     32       3 x 3/ 2    416 x 416 x   3 ->  208 x 208 x  32 0.075 BF
   1 conv     64       3 x 3/ 2    208 x 208 x  32 ->  104 x 104 x  64 0.399

AttributeError: ignored

#7. 準備自定義預訓練權重檔、資料集及相關參數檔

從 https://github.com/AlexeyAB/Darknet#how-to-train-tiny-yolo-to-detect-your-custom-objects 下載自定義預訓練權重檔 yolov4.conv.29

從Github中下載自定義資料集和相關參數檔（這個範例已幫大家準備了一份）

資料集(my_dataset.zip)中共有100張影像（＊.jpg，內容為狗臉、貓臉及人臉三種物件）及yolo格式(＊.txt)標註檔。

此資料集僅供學習測試，各影像內容版權屬原作者。建議可自行重新收集、標註，標註工具推薦使用LabelImg https://github.com/tzutalin/labelImg 進行資料標註即可，存檔時記得點選yolo格式。若產生VOC格式則須另外轉換成Yolo格式，這裡不提供轉換程式。

參數檔包含下列內容，可依資料集實際內容進行修改。

*    my_obj.data （物件資料設定，含物件類別、訓練驗證路徑、備份權重路徑）
*    my_obj.names （物件類別名稱，每一列為一個類別名稱）
*    my_yolov4-tiny-custom.cfg （設定模型組態）
*    my_train.txt （訓練內容檔案名稱，取資料集前80組，可自行定義）
*    my_valid.txt （驗證內容檔案名稱，取資料集後20組，可自行定義）

其中第3項my_yolov4-tiny-custom.cfg是從/darknet/cfg/yolov4-tiny-custom.cfg須修改而得，修改內容如下，須依自定義資料集內容而定。
*    line 6 : batch=64 # 可依顯卡記憶體調整
*    line 7 : subdivisions=1 # 可依顯卡記憶體調整
*    line 8 : width=416 # 須為32的倍數
*    line 9 : height=416 # 須為32的倍數
*    line 20 : max_batches=6000 # 類別數量(classes)x2000，目前類別數量為3
*    line 22 : steps=4800,5400 # 為max_batch值的 80%, 90%
*    line 212, 263 : filter=255 # 為(classes+5)x3
*    line 220, 269 : classes=3 # 物件類別數量

最後將相關參數檔複製到/darknet對應路徑下。  

In [ ]:
#回到使用者根目錄
%cd /content
# 下載yolov4-tiny自定義權重檔yolov4-tiny.conv.29(18.8MB)
!wget -N https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v4_pre/yolov4-tiny.conv.29
# 下載資料集和相關參數檔
!git clone https://github.com/OmniXRI/Yolov4-tiny_Colab_User_Datasets
!ls
# 相關參數到對應路徑下
!cp Yolov4-tiny_Colab_User_Datasets/my_yolov4-tiny-custom.cfg darknet/cfg/
!cp Yolov4-tiny_Colab_User_Datasets/my_obj.data darknet/data/
!cp Yolov4-tiny_Colab_User_Datasets/my_obj.names darknet/data/
!cp Yolov4-tiny_Colab_User_Datasets/my_train.txt darknet/data/
!cp Yolov4-tiny_Colab_User_Datasets/my_val.txt darknet/data/
# 將資料集解壓縮到指定路徑
!unzip Yolov4-tiny_Colab_User_Datasets/my_dataset.zip -d darknet/data
!ls darknet/data/my_dataset/

/content
--2023-10-28 10:00:43--  https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v4_pre/yolov4-tiny.conv.29
Resolving github.com (github.com)... 192.30.255.112
Connecting to github.com (github.com)|192.30.255.112|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/75388965/28807d00-3ea4-11eb-97b5-4c846ecd1d05?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20231028%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20231028T100043Z&X-Amz-Expires=300&X-Amz-Signature=f4f3367dba4bdea12979a976f5b55b39f149fb3f76dead906f98c54943d5810d&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=75388965&response-content-disposition=attachment%3B%20filename%3Dyolov4-tiny.conv.29&response-content-type=application%2Foctet-stream [following]
--2023-10-28 10:00:44--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/75388965/28807d00-3ea4-

In [ ]:
!unzip data.zip -d darknet/data
!ls darknet/data/data/

Archive:  data.zip
   creating: darknet/data/data/
 extracting: darknet/data/data/classes.txt  
  inflating: darknet/data/data/img_001.jpg  
 extracting: darknet/data/data/img_001.txt  
  inflating: darknet/data/data/img_002.jpg  
 extracting: darknet/data/data/img_002.txt  
  inflating: darknet/data/data/img_003.jpg  
 extracting: darknet/data/data/img_003.txt  
  inflating: darknet/data/data/img_004.jpg  
  inflating: darknet/data/data/img_004.txt  
  inflating: darknet/data/data/img_005.jpg  
  inflating: darknet/data/data/img_005.txt  
  inflating: darknet/data/data/img_006.jpg  
 extracting: darknet/data/data/img_006.txt  
  inflating: darknet/data/data/img_007.jpg  
 extracting: darknet/data/data/img_007.txt  
  inflating: darknet/data/data/img_008.jpg  
  inflating: darknet/data/data/img_008.txt  
  inflating: darknet/data/data/img_009.jpg  
  inflating: darknet/data/data/img_009.txt  
  inflating: darknet/data/data/img_010.jpg  
 extracting: darknet/data/data/img_010.txt  
  in

In [ ]:
!sed -i 's/my_dataset/data/' darknet/data/my_train.txt
!sed -i 's/my_data/data/' darknet/data/my_val.txt

#8. 開始訓練  
##8.1 正常重新訓練  
指定.data, .cfg, 預訓練參數檔並開始訓練，如需記錄平均精確度mAP(mean average precisions)可加上參數 -map。如果訓練過程很容易無故中止則建議移除參數 -map。  

訓練期間會記錄Loss並繪成圖表(darknet/chart_my_yolov4-tiny-custom.png)，可隨時重新開啟觀察訓練成果。  

每隔1000次會自動備份一次權重檔到雲端硬碟/my_drive/yolov4-tiny下(根據my_obj.data中backup設定值)，檔名為my_yolov4-tiny_x000.weights (x為1~n)。  

訓練期間會自動產生my_yolov4-tiny_best.weights和my_yolov4-tiny_last.weights，完成訓練會產生my_yolov4-tiny_final.weights。  

這裡要注意雲端硬體要有足夠空間存放，否則空間不足時就無法備份權重值。

In [ ]:
%cd /content/darknet
!./darknet detector train data/my_obj.data cfg/my_yolov4-tiny-custom.cfg ../yolov4-tiny.conv.29 -map

串流輸出內容已截斷至最後 5000 行。
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 37 Avg (IOU: 0.000000), count: 1, class_loss = 0.000000, iou_loss = 0.000000, total_loss = 0.000000 
 total_bbox = 55844, rewritten_bbox = 0.000000 % 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 30 Avg (IOU: 0.893819), count: 8, class_loss = 0.123682, iou_loss = 0.053342, total_loss = 0.177024 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 37 Avg (IOU: 0.000000), count: 1, class_loss = 0.000000, iou_loss = 0.000000, total_loss = 0.000000 
 total_bbox = 55852, rewritten_bbox = 0.000000 % 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 30 Avg (IOU: 0.870908), count: 8, class_loss = 0.017569, iou_loss = 0.058944, total_loss = 0.076513 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 37 Avg (IOU: 0.000000), count: 1, class_loss = 0.000000, iou_loss = 0.000000, total_loss = 0.000000 
 total_bbox = 55860, rewritten_bbox = 0

##8.2 重新接續執行訓練  
前一步驟有設定備份訓練過程的權重檔，若遇到執行到一半就斷線，此時可以從Google Drive中把最後一次權重檔 my_yolov4-tiny-custom_last.weights 複製到Colab darknet/backup下，再重新啟動訓練。

In [ ]:
!cp /my_drive/yolov4-tiny_1027/my_yolov4-tiny-custom_last.weights backup/
!ls backup/

my_yolov4-tiny-custom_last.weights


In [ ]:
!./darknet detector train data/my_obj.data cfg/my_yolov4-tiny-custom.cfg backup/my_yolov4-tiny-custom_last.weights -map -dont_show

 CUDA-version: 11080 (12000), cuDNN: 8.9.0, CUDNN_HALF=1, GPU count: 1  
 CUDNN_HALF=1 
 OpenCV version: 4.5.4
 Prepare additional network for mAP calculation...
 0 : compute_capability = 750, cudnn_half = 1, GPU: Tesla T4 
net.optimized_memory = 0 
mini_batch = 1, batch = 4, time_steps = 1, train = 0 
   layer   filters  size/strd(dil)      input                output
   0 Create CUDA-stream - 0 
 Create cudnn-handle 0 
conv     32       3 x 3/ 2    416 x 416 x   3 ->  208 x 208 x  32 0.075 BF
   1 conv     64       3 x 3/ 2    208 x 208 x  32 ->  104 x 104 x  64 0.399 BF
   2 conv     64       3 x 3/ 1    104 x 104 x  64 ->  104 x 104 x  64 0.797 BF
   3 route  2 		                       1/2 ->  104 x 104 x  32 
   4 conv     32       3 x 3/ 1    104 x 104 x  32 ->  104 x 104 x  32 0.199 BF
   5 conv     32       3 x 3/ 1    104 x 104 x  32 ->  104 x 104 x  32 0.199 BF
   6 route  5 4 	                           ->  104 x 104 x  64 
   7 conv     64       1 x 1/ 1    104 x 104 x  64 

#9. 測試訓練結果  
1.   完成訓練後，可將結果權重檔my_yolov4-tiny-custom_final.weights複製到Colab虛擬機中。
2.   執行darknet物件偵測推論，指定 data, config, weight, image 所在位置，/Yolov4-tiny_Colab_User_Datasets下有test01.jpg和test02.jpg可供測試。
3.   顯示結果影像predictions.jpg

In [ ]:
!cp /my_drive/yolov4-tiny_1027/my_yolov4-tiny-custom_best.weights ../

!./darknet detector train data/my_obj.data cfg/my_yolov4-tiny-custom.cfg ../my_yolov4-tiny-custom_best.weights ../img_500.jpg
import cv2
from google.colab.patches import cv2_imshow

imgResult = cv2.imread('predictions.jpg')
cv2_imshow(imgResult)

 CUDA-version: 11080 (12000), cuDNN: 8.9.0, CUDNN_HALF=1, GPU count: 1  
 CUDNN_HALF=1 
 OpenCV version: 4.5.4
my_yolov4-tiny-custom
 0 : compute_capability = 750, cudnn_half = 1, GPU: Tesla T4 
net.optimized_memory = 0 
mini_batch = 8, batch = 32, time_steps = 1, train = 1 
   layer   filters  size/strd(dil)      input                output
   0 Create CUDA-stream - 0 
 Create cudnn-handle 0 
conv     32       3 x 3/ 2    416 x 416 x   3 ->  208 x 208 x  32 0.075 BF
   1 conv     64       3 x 3/ 2    208 x 208 x  32 ->  104 x 104 x  64 0.399 BF
   2 conv     64       3 x 3/ 1    104 x 104 x  64 ->  104 x 104 x  64 0.797 BF
   3 route  2 		                       1/2 ->  104 x 104 x  32 
   4 conv     32       3 x 3/ 1    104 x 104 x  32 ->  104 x 104 x  32 0.199 BF
   5 conv     32       3 x 3/ 1    104 x 104 x  32 ->  104 x 104 x  32 0.199 BF
   6 route  5 4 	                           ->  104 x 104 x  64 
   7 conv     64       1 x 1/ 1    104 x 104 x  64 ->  104 x 104 x  64 0.089 BF

AttributeError: ignored